sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/PubsubGrpcClient.java

```diff
@@ -0,0 +1,390 @@
+      @Nullable String timestampLabel, @Nullable String idLabel,
+      GcpOptions options) throws IOException {
+    ManagedChannel channel = NettyChannelBuilder
+        .forAddress(PUBSUB_ADDRESS, PUBSUB_PORT)
+        .negotiationType(NegotiationType.TLS)
+        .sslContext(GrpcSslContexts.forClient().ciphers(null).build())
+        .build();
+    // TODO: GcpOptions needs to support building com.google.auth.oauth2.Credentials from the
+    // various command line options. It currently only supports the older
+    // com.google.api.client.auth.oauth2.Credentials.
+    GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
+    return new PubsubGrpcClient(timestampLabel, idLabel, channel, credentials);
+  }
+
+  /**
+   * Gracefully close the underlying netty channel.
+   */
+  @Override
+  public void close() {
+    Preconditions.checkState(publisherChannel != null, "Client has already been closed");
+    publisherChannel.shutdown();
+    try {
+      publisherChannel.awaitTermination(TIMEOUT_S, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      // Ignore.
+      Thread.currentThread().interrupt();
+    }
+    publisherChannel = null;
+    cachedPublisherStub = null;
+    cachedSubscriberStub = null;
+  }
+
+  /**
+   * Return channel with interceptor for returning credentials.
+   */
+  private Channel newChannel() throws IOException {
+    Preconditions.checkState(publisherChannel != null, "PubsubGrpcClient has been closed");
+    ClientAuthInterceptor interceptor =
+        new ClientAuthInterceptor(credentials, Executors.newSingleThreadExecutor());
+    return ClientInterceptors.intercept(publisherChannel, interceptor);
+  }
+
+  /**
+   * Return a stub for making a publish request with a timeout.
+   */
+  private PublisherGrpc.PublisherBlockingStub publisherStub() throws IOException {
+    if (cachedPublisherStub == null) {
+      cachedPublisherStub = PublisherGrpc.newBlockingStub(newChannel());
+    }
+    return cachedPublisherStub.withDeadlineAfter(TIMEOUT_S, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Return a stub for making a subscribe request with a timeout.
+   */
+  private SubscriberGrpc.SubscriberBlockingStub subscriberStub() throws IOException {
+    if (cachedSubscriberStub == null) {
+      cachedSubscriberStub = SubscriberGrpc.newBlockingStub(newChannel());
+    }
+    return cachedSubscriberStub.withDeadlineAfter(TIMEOUT_S, TimeUnit.SECONDS);
+  }
+
+  @Override
+  public int publish(String topic, Iterable<OutgoingMessage> outgoingMessages) throws IOException {
+    PublishRequest.Builder request = PublishRequest.newBuilder()
+                                                   .setTopic(topic);
+    for (OutgoingMessage outgoingMessage : outgoingMessages) {
+      PubsubMessage.Builder message =
+          PubsubMessage.newBuilder()
+                       .setData(ByteString.copyFrom(outgoingMessage.elementBytes));
+
+      if (timestampLabel != null) {
```

Comment:
In both the source and the sink I found I needed my own message representation. When it came to making this impl public I was forced to avoid any PubsubMessages in the API. By dealing with timestampLabel (and idLabel) here I get to kill two birds as it were: have the intermediate representation I need and solve the API problem.

Remember, every message goes through three representations (T, OutgoingMessage, PubsubMessage). That's already pretty unfortunate.